internal/envoy/v3/cluster.go

```diff
@@ -162,6 +162,48 @@ func ExtensionCluster(ext *dag.ExtensionCluster) *envoy_cluster_v3.Cluster {
 	return cluster
 }
 
+// DNSNameCluster builds a envoy_cluster_v3.Cluster for the given *dag.DNSNameCluster.
+func DNSNameCluster(c *dag.DNSNameCluster) *envoy_cluster_v3.Cluster {
+	cluster := clusterDefaults()
+
+	cluster.Name = envoy.DNSNameClusterName(c)
+	cluster.DnsLookupFamily = parseDNSLookupFamily(c.DNSLookupFamily)
+	cluster.ClusterDiscoveryType = &envoy_cluster_v3.Cluster_Type{
+		Type: envoy_cluster_v3.Cluster_STRICT_DNS,
+	}
+
+	var port int
+	var transportSocket *envoy_core_v3.TransportSocket
+	switch c.Scheme {
+	case "http":
+		port = 80
+	case "https":
+		port = 443
+		// TODO TLS validation
+		transportSocket = UpstreamTLSTransportSocket(UpstreamTLSContext(nil, c.Address, nil))
+	}
+
+	cluster.LoadAssignment = &envoy_endpoint_v3.ClusterLoadAssignment{
```

Comment:
nit: sort of duplicated with the `StaticClusterLoadAssignment` helper, which does take a dag.Service rather than a host/port but maybe could be de-duped?

Reply:
also could maybe use the `Endpoints` helper?

Reply:
Yeah, good point, let me take another look at this -- initially I didn't want to start changing the existing helpers since they were pretty tied to the dag.Service, but there's probably some opportunity for reuse.

Reply:
Take a look at https://github.com/projectcontour/contour/pull/4723/commits/1e1810cb37630d1806c12ef5ea1af94afe7096c5, does some refactoring across this area.